rig/bitfield.py

```diff
@@ -361,15 +364,49 @@ def assign_fields(self):
         Users should typically call this method after all field values have
         been assigned, otherwise fields may be fixed at an inadequate size.
         """
-        # We must fix fields at every level of the heirarchy sepeartely
+        class Node(namedtuple("Node", "bitfield children")):
+            def recurse_assign_unfixed_fields(self, assigned_bits=0,
+                                              excluded_fields=set()):
+                # Call this for the children first
+                excludes = set(i for (i, f) in self.bitfield._enabled_fields())
+                new_assigned_bits = assigned_bits
+                for child in self.children:
+                    # Children can assign bits independently
+                    new_assigned_bits |= child.recurse_assign_unfixed_fields(
+                        assigned_bits, excludes)
+
+                # Now assign locally
+                assigned_bits =\
+                    self.bitfield._assign_enabled_fields_without_fixed_start(
+                        new_assigned_bits, excluded_fields
+                    )
+
+                # Check that everything HAS been assigned
+                for i, f in self.bitfield._enabled_fields():
+                    if i not in excluded_fields:
+                        assert f.length is not None and f.start_at is not None
+
+                return assigned_bits
+
+        # We must fix fields at every level of the hierarchy separately
         # (otherwise fields of children won't be allowed to overlap). Here we
-        # do a breadth-first iteration over the heirarchy, fixing the fields at
-        # each level.
-        unsearched_heirarchy = [BitField(self.length, self.fields)]
-        while unsearched_heirarchy:
-            ks = unsearched_heirarchy.pop(0)
-            ks._assign_enabled_fields()
-            # Look for potential children in the herarchy
+        # do a breadth-first iteration over the hierarchy to fix fields with
+        # given starting positions; then we do depth-first to fix other fields.
```

Comment:
Oh cool: those line refs refer to the actual line not just the line number! Clever github!